### PR TITLE
Mutual Friendships Implementation

### PR DESCRIPTION
**Pull request by acushlakoncept**
This pull request contains the implementation of mutual friends.
We created a method in the user model that filters out mutual friends between two users.

### CODE REVIEW

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello @acushlakoncept and @somoye123 
### Nice job you have done so far :clap: However, you is conflict in your branch `mutual-friends` fix this and submit for another code review.
merge

- Good luck and happy coding!

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great Work @team :clap: :clap: However, some requirements need to be implemented in order to get approval for this milestone :arrow_down: 

##### Behaviour
* [x] There is a button to invite user to friendship on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
##### Associations

* [x] when friendship is confirmed, an additional row with reversed user-friend pair is created

```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

* [x] thanks to this, we can find user friends only by user_id in friendships table

```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

* [x] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

* [x] User#friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```

* [x] in order to get all posts written by user on their friends, students can use simple method in User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
##### Tests
* [x] All Units should be tested, plus, Integration tests are added for the main features of the app (use of capybara gem is advised)

### That's it :relieved: When finished, kindly submit a code re-review request :pray: 
Happy coding :v: 
**_Ahmed Mahfoudh_**

**Comment by acushlakoncept** [comment]
> ### Great Work @team 👏 👏 However, some requirements need to be implemented in order to get approval for this milestone ⬇️
> ##### Behaviour
> * [x]  There is a button to invite user to friendship on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
> 
> ##### Associations
> * [x]  when friendship is confirmed, an additional row with reversed user-friend pair is created
> 
> ```ruby
> # Friendship model
> 
> def confirm_friend
>   self.update_attributes(confirmed: true)
>   Friendship.create!(friend_id: self.user_id,
>                   user_id: self.friend_id,
>                   confirmed: true)
> end
> ```
> 
> * [x]  thanks to this, we can find user friends only by user_id in friendships table
> 
> ```ruby
> # User model
> 
> has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
> has_many :friends, through: :confirmed_friendships
> ```
> 
> * [x]  User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
> 
> ```ruby
> # User model
> 
> # Users who needs to confirm friendship
> has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
> has_many :pending_friends, through: :pending_friendships, source: :friend
> ```
> 
> * [x]  User#friend_requests can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
> 
> ```ruby
> # User model
> 
> # Users who requested to be friends (needed for notifications)
> has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
> has_many :friend_requests, through: :inverted_friendships
> ```
> 
> * [x]  in order to get all posts written by user on their friends, students can use simple method in User model (in order to NOT do it with Enumerables)
> 
> ```ruby
> # User model
> 
> def friends_and_own_posts
>   Post.where(user: (self.friends + self))
>   # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
> end
> ```
> 
> ##### Tests
> * [x]  All Units should be tested, plus, Integration tests are added for the main features of the app (use of capybara gem is advised)
> 
> ### That's it 😌 When finished, kindly submit a code re-review request 🙏
> Happy coding ✌️
> **_Ahmed Mahfoudh_**

Thank you @stratospherique, for all the heads up and refactoring possibilities using Active Record.

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -7,6 +7,31 @@ def index
 
   def show
     @user = User.find(params[:id])
+
     @posts = @user.posts.ordered_by_most_recent
+
+    # @pending = @user.pending_friends
+    # @incoming = @user.friend_requests
```
Comment (kobiyoyo) [inline app/controllers/users_controller.rb]: Lets delete this commented code

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team     
#### Great Job so far, everything works fine, but there are a few more things required. 
### General
- [x] Lets write integration test for this project, using Rspec and Capybara.
- [x] Check comments below  

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by uryelah — APPROVED** [review]
Hi **Somoye Somoye & Uduak Essien**, I'm Sarah your reviewer for this PR 👩‍🚀 

Great job adding the capybara testing to your project! That will sure look impressive in your CV skills.

Go forth and rock your next milestone!  👨‍🎤 🎸  

-------

## **Your project is approved! 🎊 🥅 💛**

**You may close this Pull Request and merge the changes**


---------

    
    
**Keep being awesome devs and good luck with your next project!**

---------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Comment by acushlakoncept** [comment]
> Hi **Somoye Somoye & Uduak Essien**, I'm Sarah your reviewer for this PR 👩‍🚀
> 
> Great job adding the capybara testing to your project! That will sure look impressive in your CV skills.
> 
> Go forth and rock your next milestone! 👨‍🎤 🎸
> 
> ## **Your project is approved! 🎊 🥅 💛**
> **You may close this Pull Request and merge the changes**
> 
> **Keep being awesome devs and good luck with your next project!**
> 
> Cheers and Happy coding!👏👏👏
> 
> Code review by **Sarah Uryelah Silva Chamorro**
> 
> (Feel free to contact me on slack by the handle of **sarah** if you have any additional question)

Thank you @uryelah

**Pull request merged**